File: lesson14/MultiTouchView/app/src/main/java/com/chinalwb/multitouchview/view/MultiTouchView3.java
package com.chinalwb.multitouchview.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.util.AttributeSet;
import android.util.Log;
import android.util.SparseArray;
import android.view.MotionEvent;
import android.view.View;

import com.chinalwb.multitouchview.Utils;

import java.net.PasswordAuthentication;

import androidx.annotation.Nullable;

public class MultiTouchView3 extends View {

    private Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private SparseArray<Path> paths = new SparseArray<>(5);

    public MultiTouchView3(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    {
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeWidth(Utils.dp2px(5));
        paint.setColor(Color.RED);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setStrokeJoin(Paint.Join.ROUND);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        for (int i = 0; i < paths.size(); i++) {
            canvas.drawPath(paths.valueAt(i), paint);
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        int actionIndex = event.getActionIndex();
        int pointerId = event.getPointerId(actionIndex);
        Log.e("XX", "pointer id == " + pointerId);

        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
            case MotionEvent.ACTION_POINTER_DOWN:
                Path path = new Path();
                path.moveTo(event.getX(actionIndex), event.getY(actionIndex));
                paths.put(pointerId, path);
                Log.e("XX", "paths size == " + paths.size());
                break;
            case MotionEvent.ACTION_MOVE:
                for (int i = 0; i < event.getPointerCount(); i++) {
                    pointerId = event.getPointerId(i);
                    Path pointerPath = paths.get(pointerId);
                    pointerPath.lineTo(event.getX(i), event.getY(i));
                }
                invalidate();
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_POINTER_UP:
                path = paths.get(pointerId);
                if (null != path) {
                    paths.remove(pointerId);
                    invalidate();
                }
                break;
        }
        return true;
    }
}

File: lesson14/MultiTouchView/app/src/main/java/com/chinalwb/multitouchview/view2/MultiTouchView3.kt
package com.chinalwb.multitouchview.view2

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.util.AttributeSet
import android.util.SparseArray
import android.view.MotionEvent
import android.view.View
import com.chinalwb.multitouchview.Utils

class MultiTouchView3 (context: Context, attributeSet: AttributeSet) : View(context, attributeSet) {

    private var paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var path = Path()
    private var pathSparseArray = SparseArray<Path>()

    init {
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = Utils.dp2px(2)
        paint.strokeCap = Paint.Cap.ROUND
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        for (i in 0 until pathSparseArray.size()) {
            var path = pathSparseArray.valueAt(i)
            canvas!!.drawPath(path, paint)
        }
        canvas!!.drawPath(path, paint)
    }

    override fun onTouchEvent(event: MotionEvent?): Boolean {
        var pointerId = event!!.getPointerId(event.actionIndex)
        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN,
            MotionEvent.ACTION_POINTER_DOWN -> {
                var newPath = Path()
                pathSparseArray.append(pointerId, newPath)
                newPath.moveTo(event.getX(event.actionIndex), event.getY(event.actionIndex))
            }
            MotionEvent.ACTION_MOVE -> {
                for (i in 0 until event.pointerCount) {
                    var currentPointerId = event.getPointerId(i)
                    var cPath = pathSparseArray.get(currentPointerId)
                    cPath.lineTo(event.getX(i), event.getY(i))
                }
                invalidate()
            }
            MotionEvent.ACTION_UP,
            MotionEvent.ACTION_POINTER_UP-> {
                pathSparseArray.remove(pointerId)
                invalidate()
            }
        }
        return true
    }
}